File: src/javasynchronized/ProducerConsumer.java
package javasynchronized;

import java.util.LinkedList;
import java.util.Scanner;

public class ProducerConsumer {
    private final LinkedList<String> messages;
    public ProducerConsumer() {
        messages = new LinkedList<>();
    }

    public synchronized void addMessage(String message) {
        messages.addLast(message);
        notifyAll();
    }
    public synchronized String getMessage() throws InterruptedException {
        while (messages.isEmpty()) wait();
        return messages.removeFirst();
    }

    public static void main(String[] args) {
        ProducerConsumer pc = new ProducerConsumer();
        Scanner scanner = new Scanner(System.in);

        // Producer
        new Thread(() -> {
            while (true) {
                pc.addMessage(scanner.nextLine().trim());
            }
        }).start();

        // Consumer
        new Thread(() -> {
            while (true) {
                try {
                    System.out.println("got: " + pc.getMessage());
                    Thread.sleep(1000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }).start();
    }
}

File: src/SemaphoresTutorial/ProducerConsumer.kt
package SemaphoresTutorial

import java.util.*
import java.util.concurrent.Semaphore

class BlockingQueue {

}
object ProducerConsumer {
    @JvmStatic
    fun main(args: Array<String>) {
        val signal = Semaphore(0)
        val mutex = Semaphore(1);
        val list = LinkedList<String>()
        var number = 0;
        val scanner = Scanner(System.`in`)
        Thread {
            while (true) {
                val scanned = scanner.nextLine().trim()

                mutex.acquire()
                list.addLast(scanned)
                mutex.release()

                signal.release()
            }
        }.start()

        Thread {
            while (true) {
                signal.acquire()

                mutex.acquire()
                val first = list.removeFirst()
                mutex.release()

                println("got $first")
                Thread.sleep(3000)
            }
        }.start()
    }
}